src/Microsoft.AspNetCore.Hosting/Internal/ApplicationLifetime.cs

```diff
@@ -115,9 +102,32 @@ public void NotifyStopped()
             }
         }
 
-        private void ExecuteHandlers(Action<IApplicationLifetimeEvents> callback)
+        private void ExecuteHandlers(CancellationTokenSource cancel, Action<IApplicationLifetimeEvents> callback)
         {
+            // Noop if this is already cancelled
+            if (cancel.IsCancellationRequested)
+            {
+                return;
+            }
+
             List<Exception> exceptions = null;
+
+            try
+            {
+                // Run the cancellation token callbacks
+                cancel.Cancel(throwOnFirstException: false);
+            }
+            catch (Exception ex)
+            {
+                if (exceptions == null)
```

Comment:
nit: this code does not run often, we can handle 3 lists per life of application.